### PR TITLE
Implement QUICHE quic_text_utils_impl

### PR DESCRIPTION
**Pull request by mpwarres**
*Description*:
Implement QUICHE quic_text_utils_impl.
Also implement part of quic_test_impl, which is needed to test quic_text_utils_impl.
Tested via `bazel test @com_googlesource_quiche//:quic_platform_test`

For #2557.

Signed-off-by: Michael Warres <mpw@google.com>

*Risk Level*: low (not yet used in Envoy)
*Testing*: unit tests
*Docs Changes*: none
*Release Notes*: none

### CODE REVIEW

**Comment by danzh2010** [comment]
LGTM

File: source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    return absl::StartsWith(data, prefix);
+  }
+
+  static bool EndsWithIgnoreCase(QuicStringPiece data, QuicStringPiece suffix) {
+    return absl::EndsWithIgnoreCase(data, suffix);
+  }
+
+  static QuicString ToLower(QuicStringPiece data) { return absl::AsciiStrToLower(data); }
+
+  static void RemoveLeadingAndTrailingWhitespace(QuicStringPiece* data) {
```
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h]: Can we use absl::StripAsciiWhitespace given that QuicStringPiece is string_view under the hood?

File: source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  static QuicString Uint64ToString(uint64_t in) { return absl::StrCat(in); }
+
+  static QuicString HexEncode(QuicStringPiece data) { return absl::BytesToHexString(data); }
+
+  static QuicString Hex(uint32_t v) { return absl::StrCat(absl::Hex(v)); }
+
+  static QuicString HexDecode(QuicStringPiece data) { return absl::HexStringToBytes(data); }
+
+  static void Base64Encode(const uint8_t* data, size_t data_len, QuicString* output) {
+    absl::Base64Escape(QuicString(reinterpret_cast<const char*>(data), data_len), output);
```
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h]: WebSafeBase64Escape seems to not do padding but does have slightly different rules.  Is it usable or no?

File: source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h (hunk 1 above)
Comment (mpwarres) [reply source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h]: Oh! Yes, definitely. Somehow I missed that method when looking for one to use, thanks for pointing it out.

File: source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h (hunk 2 above)
Comment (mpwarres) [reply source/extensions/quic_listeners/quiche/platform/quic_text_utils_impl.h]: I'm not sure it's safe to: QuicTextUtils::Base64Encode is used for generating client channel IDs, so the minor difference in characters used could actually have an effect. (That said, this is for channel ID, which last I heard is being (or maybe has been) deprecated.)

Ok to leave as-is for the time being, and then hopefully remove once channel ID code is removed from QUICHE core code?

**Comment by mpwarres** [comment]
@alyssawilk I think this is all set to be merged, thanks!

**Pull request merged**